src/python/pants/backend/codegen/protobuf/scala/symbol_mapper.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import os
+import re
+from collections import defaultdict
+from typing import DefaultDict, Mapping
+
+from pants.backend.codegen.protobuf.target_types import AllProtobufTargets, ProtobufSourceField
+from pants.engine.addresses import Address
+from pants.engine.fs import Digest, DigestContents, FileContent
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import HydratedSources, HydrateSourcesRequest
+from pants.engine.unions import UnionRule
+from pants.jvm.dependency_inference import symbol_mapper
+from pants.jvm.dependency_inference.artifact_mapper import MutableTrieNode
+from pants.jvm.dependency_inference.symbol_mapper import FirstPartyMappingRequest, SymbolMap
+from pants.jvm.subsystems import JvmSubsystem
+from pants.jvm.target_types import JvmResolveField
+from pants.util.ordered_set import OrderedSet
+
+_ResolveName = str
+
+
+class FirstPartyProtobufScalaTargetsMappingRequest(FirstPartyMappingRequest):
+    pass
+
+
+# Determine generated Java/Scala package name
+# * https://scalapb.github.io/docs/generated-code
+def _determing_package_name(file: FileContent) -> str:
+    base_name, _, _ = os.path.basename(file.path).partition(".")
+    package_definition = _parse_package_definition(file.content)
+    return f"{package_definition}.{base_name}" if package_definition else base_name
+
+
+_QUOTE_CHAR = r"(?:'|\")"
+_JAVA_PACKAGE_OPTION_RE = re.compile(
+    rf"^\s*option\s+java_package\s+=\s+{_QUOTE_CHAR}(.+){_QUOTE_CHAR};"
+)
+_PACKAGE_RE = re.compile(r"^\s*package\s+(.+);")
+
+
+def _parse_package_definition(content_raw: bytes) -> str | None:
+    content = content_raw.decode()
+    for line in content.splitlines():
+        m = _JAVA_PACKAGE_OPTION_RE.match(line)
+        if m:
+            return m.group(1)
+        m = _PACKAGE_RE.match(line)
+        if m:
+            return m.group(1)
+    return None
```

Comment:
It would be good to lift both of these into a shared `protobuf` package somewhere for use by all of the protobuf producing languages. Or perhaps even the entire `map_first_party_protobuf_targets_to_symbols` function... although that might be more tricky.

I think that we might also already have dependency inference for python protobuf, so you might see if there is some reuse possible there too?

Reply:
yeah, lots of repetition between the two implementations, I'll come up with something similar as what I did for the Thrift one. Regarding Python, it seems the approach is quite different there, guessing final package name from the file path instead of parsing it from the contents.